src/olympusmn_int.cpp

```diff
@@ -1391,67 +1436,46 @@ std::ostream& OlympusMakerNote::printCs0x0301(std::ostream& os, const Value& val
-      {{0, 0}, N_("Off")},
-      {{0, 1280}, N_("Off")},
-      {{1, 1280}, N_("Soft Focus")},
-      {{2, 1280}, N_("Pop Art")},
-      {{3, 1280}, N_("Pale & Light Color")},
-      {{4, 1280}, N_("Light Tone")},
-      {{5, 1280}, N_("Pin Hole")},
-      {{6, 1280}, N_("Grainy Film")},
-      {{9, 1280}, N_("Diorama")},
-      {{10, 1280}, N_("Cross Process")},
-      {{12, 1280}, N_("Fish Eye")},
-      {{13, 1280}, N_("Drawing")},
-      {{14, 1280}, N_("Gentle Sepia")},
-      {{15, 1280}, N_("Pale & Light Color II")},
-      {{16, 1280}, N_("Pop Art II")},
-      {{17, 1280}, N_("Pin Hole II")},
-      {{18, 1280}, N_("Pin Hole III")},
-      {{19, 1280}, N_("Grainy Film II")},
-      {{20, 1280}, N_("Dramatic Tone")},
-      {{21, 1280}, N_("Punk")},
-      {{22, 1280}, N_("Soft Focus 2")},
-      {{23, 1280}, N_("Sparkle")},
-      {{24, 1280}, N_("Watercolor")},
-      {{25, 1280}, N_("Key Line")},
-      {{26, 1280}, N_("Key Line II")},
-      {{27, 1280}, N_("Miniature")},
-      {{28, 1280}, N_("Reflection")},
-      {{29, 1280}, N_("Fragmented")},
-      {{31, 1280}, N_("Cross Process II")},
-      {{32, 1280}, N_("Dramatic Tone II")},
-      {{33, 1280}, N_("Watercolor I")},
-      {{34, 1280}, N_("Watercolor II")},
-      {{35, 1280}, N_("Diorama II")},
-      {{36, 1280}, N_("Vintage")},
-      {{37, 1280}, N_("Vintage II")},
-      {{38, 1280}, N_("Vintage III")},
-      {{39, 1280}, N_("Partial Color")},
-      {{40, 1280}, N_("Partial Color II")},
-      {{41, 1280}, N_("Partial Color III")},
-  };
-
+std::ostream& OlympusMakerNote::print0x0529(std::ostream& os, const Value& value, const ExifData* metadata) {
   if (value.count() != 4 || value.typeId() != unsignedShort) {
-    return os << value;
+    return os << "(" << value << ")";
   }
 
-  auto v0 = static_cast<uint16_t>(value.toInt64(0));
-  auto v1 = static_cast<uint16_t>(value.toInt64(1));
+  const auto v0 = value.toInt64(0);
 
-  for (auto&& filter : artFilters) {
-    if (filter.val[0] == v0 && filter.val[1] == v1) {
-      return os << filter.label;
-    }
+  printTag<std::size(artFilters), artFilters>(os, v0, metadata);
+
+  if (v0 == 39) {  // The "Partial color" option also has a color choice
+    const auto v3 = value.toInt64(3);
+    os << " (" << _("position") << " " << (v3 + 1) << ")";
+    return os;
   }
-  return os << "";
+
+  return os;
 }  // OlympusMakerNote::print0x0529
 
+//! OlympusCs ArtFilter, tag 0x0604, OlympusCs ImageStabilization
+std::ostream& OlympusMakerNote::print0x0604(std::ostream& os, const Value& value, const ExifData* metadata) {
+  if (value.count() != 1 || value.typeId() != unsignedLong) {
+    return os << "(" << value << ")";
+  }
+
+  const auto v0 = value.toInt64(0);
+
+  if (metadata) {
+    const auto pos = metadata->findKey(ExifKey("Exif.Image.Model"));
+    if (pos != metadata->end() && pos->count() != 0) {
+      const auto model = pos->toString();
+      if (model.find("E-M10MarkII") != std::string::npos) {
```

Comment:
@kmilos:
> Are we sure we don't want to just go ahead and replace the existing mode1/2/3/4 with these more specific labels for all Olympus bodies perhaps?

I looked at a couple of Olympus instruction manuals but I wasn't sure if the values were the same for all models. ExifTool only uses the generic labels.

I thought that `print0x0604()` could be extended in the future for other Olympus models when we are requested. However, if you think it is better to rename the labels then I will make the change.

Reply:
I'm not sure... Maintaining this model per model is not ideal since it's source code and not some external xml/json database...

Is there some other hint (e.g. MakerNote or some other field version) that can help? What do Olympus users say?

Reply:
@kmilos:
> Are we sure we don't want to just go ahead and replace the existing mode1/2/3/4 with these more specific labels for all Olympus bodies perhaps?

> Is there some other hint (e.g. MakerNote or some other field version) that can help? What do Olympus users say?

I took your recommendation and requested sample photos on https://discuss.pixls.us/t/request-for-photos-olympus-image-stabilisation-metadata/31240 . The sample files allowed me to replace the existing 1/2/3/4 modes with the new labels.